### ISSUE TITLE
Support for JWT Header TYP as "at+jwt"

### ISSUE THREAD

**Issue by tekgainers** [spring-projects/spring-security#9900]
**Expected Behavior**

Currently, if the JWT is having typ as "at+jwt", the token is rejected with message "Failed to authenticate since the JWT was invalid". Spring Security Oauth2 Resource Server with JWT as bearer token should accept typ as "at+jwt" as well.

**Current Behavior**

Currently, if the JWT is having typ as "at+jwt", the token is rejected with message "Failed to authenticate since the JWT was invalid".

My Authorization server is issuing JWT access token with typ as "at+jwt" as per the following draft: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-access-token-jwt-11

How has this issue affected you? JWT token is rejected although this is correct as per Authorization server
What are you trying to accomplish? Validate JWT Bearer token using Spring Security OAuth2 Resource Server capabilities.
What other alternatives have you considered? At this point, in the documentation I didn;t find a way where I can customize the JWT Validator. Although, documentation is around our own Audience or other validations once the JWT is accepted by Spring Security.
Are you aware of any workarounds? Not at this point of time.

**Comment by tekgainers** [comment]
@jzheaux can you please have a look if this is bug or this needs enhancement in library?

Thanks in advance.

**Comment by jzheaux** [comment]
The reason for this behavior is Nimbus defaults to checking that the `typ` is `JWT`. I am not certain whether Nimbus yet supports [the `at+jwt` draft](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-access-token-jwt-11#section-4) - Spring Security would likely rely on Nimbus for this kind of support.

To get the resource server to accept `"typ" : "at+jwt"`, you can configure Nimbus like so:

```java
@Bean 
JwtDecoder jwtDecoder() {
	DefaultJOSEObjectTypeVerifier<SecurityContext> verifier =
		new DefaultJOSEObjectTypeVerifier<>(new JOSEObjectType("at+jwt"));
	NimbusJwtDecoder decoder = NimbusJwtDecoder.withJwkSetUri(this.uri)
		.jwtProcessorCustomizer((processor) -> processor.setJWSTypeVerifier(verifier))
		.build()
        // ... any other decoder settings
	return decoder;
}
```

For passivity reasons, I imagine that Spring Security will continue to use the Nimbus defaults. Once the draft gets finalized, the may be an opportunity to make this configuration simpler.

**Comment by ymajoros** [comment]
Thanks @jzheaux .
What about this pull request, which solves the problem IMO?
https://github.com/spring-projects/spring-security/pull/13186

**Comment by jzheaux** [comment]
@ymajoros, I don't think that it shoud be in `JwtDecoders`, but `NimbusJwtDecoder`. Also, it is missing the needed validation. Please see [my comment](https://github.com/spring-projects/spring-security/issues/9900#issuecomment-1555274171) about what needs to be done to add this support.

**Comment by ymajoros** [comment]
After spending some time anymore, I'm not convinced we can do more validation for now, according to RFC 9068 (at+jwt) and RFC 7519 (jwt):

- if we strictly follow RFC 9068, we should only accept application/at+jwt or at+jwt (the latter being recommended)
- besides the `typ` check already checked in this pull request, we should check:
   -  `iss`, which we already do
   - signature, which we already do
   - exp, which we already do
   - `aud`, which we don't, but I suggest to keep this as a stricter option for later, as we should also check it for plain jwt anyway, according to RFC 7519

So, I think the current implementation is enough. I'd just add the optional application/ prefix.

**Comment by ymajoros** [comment]
@shartte true, but that seems out of scope for this issue and a different feature that should also be implemented, isn't it?

**Comment by ymajoros** [comment]
@jzheaux can you look at it again in light of my previous comment?

**Comment by burl21** [comment]
Is there any update on "at+jwt" support ?

**Comment by daniel-susumu** [comment]
Is there any recommended workaround?

**Comment by BilalKais** [comment]
Any updates on this one @jzheaux?

We've used the following workaround in our application:

`
  @Bean
  JwkSetUriJwtDecoderBuilderCustomizer jwkSetUriJwtDecoderBuilderCustomizer() {
    final var verifier = new DefaultJOSEObjectTypeVerifier<>(new JOSEObjectType("at+jwt"));
    return builder -> builder.jwtProcessorCustomizer((processor) -> processor.setJWSTypeVerifier(verifier));
  }
`

This approach works well when using either `resourceserver.jwt.issuer-uri` or `resourceserver.jwt.jwk-set-uri`, both of which rely on the `jwkSetUriJwtDecoderBuilder`. However, when using `public-key-location`, we don't have a way to customize the JWT processor without creating a completely new decoder. This is problematic as it leads to losing access to features like the audience validator, which would otherwise be provided.

**Comment by jzheaux** [comment]
Since there is a PR, I'll close this ticket in favor of that and we can continue the conversation over there: #13186
